optimism/src/keccak/witness.rs

```diff
@@ -83,15 +83,76 @@ impl<Fp: Field> KeccakInterpreter for KeccakEnv<Fp> {
         self.write_column(KeccakColumn::FlagRound, round);
     }
 
-    fn run_sponge(&mut self, _sponge: Sponge) {
-        todo!()
-    }
-    fn run_absorb(&mut self, _absorb: Absorb) {
-        todo!()
+    fn run_sponge(&mut self, sponge: Sponge) {
+        match sponge {
+            Sponge::Absorb(absorb) => self.run_absorb(absorb),
+            Sponge::Squeeze => self.run_squeeze(),
+        }
     }
+
     fn run_squeeze(&mut self) {
-        todo!()
+        self.write_column(KeccakColumn::FlagSqueeze, 1);
+
+        // Compute witness values
+        let state = self.prev_block.clone();
+        let shifts = Keccak::shift(&state);
+        let dense = Keccak::collapse(&Keccak::reset(&shifts));
+        let bytes = Keccak::bytestring(&dense);
+
+        // Write squeeze-related columns
+        for (i, value) in state.iter().enumerate() {
+            self.write_column(KeccakColumn::SpongeOldState(i), *value);
+        }
+        for (i, value) in bytes.iter().enumerate().take(HASH_BYTELENGTH) {
+            self.write_column(KeccakColumn::SpongeBytes(i), *value);
+        }
+        for (i, value) in shifts.iter().enumerate().take(QUARTERS * WORDS_IN_HASH) {
+            self.write_column(KeccakColumn::SpongeShifts(i), *value);
+        }
+
+        // Rest is zero thanks to null_state
+
+        // TODO: more updates to the env?
+    }
+
+    fn run_absorb(&mut self, absorb: Absorb) {
+        self.set_flag_absorb(absorb);
+
+        // Compute witness values
+        let ini_idx = self.block_idx * RATE_IN_BYTES;
```

Comment:
The drawing I did during our call. Only if you want to add it.
```
        //    Permutation + Mode of Operation (Sponge)
        //    state -> permutation(state) -> state'
        //              ----> either [0] or state'
        //             |            new state = Exp(block)
        //             |         ------------------------
        //    Absorb: state  + [  block      +     0...0 ]
        //                       1088 bits          512
        //            ----------------------------------
        //                         XOR STATE
```

Reply:
Nit (not blocking, and only a personal taste): I would simply reorganize the variable definition in a different order
```
        let ini_idx = self.block_idx * RATE_IN_BYTES;
        let mut block = self.padded[ini_idx..ini_idx + RATE_IN_BYTES].to_vec();
        // Pad with zeros
        block.append(&mut vec![0; CAPACITY_IN_BYTES]);

        let old_state = self.prev_block.clone();
        let new_state = Keccak::expand_state(&block);
        let xor_state = old_state
            .iter()
            .zip(new_state.clone())
            .map(|(x, y)| x + y)
            .collect::<Vec<u64>>();
```

Reply:
I included them in https://github.com/o1-labs/proof-systems/pull/1559/